zerver/lib/remote_server.py

```diff
@@ -245,6 +250,29 @@ def send_json_to_push_bouncer(
     )
 
 
+PUSH_NOTIFICATIONS_RECENTLY_WORKING_REDIS_KEY = "push_notifications_recently_working_ts"
+
+
+def record_push_notifications_recently_working() -> None:
+    # Record the timestamp in redisp, marking that push notifications
+    # were working as of this moment.
+
+    redis_key = redis_utils.REDIS_KEY_PREFIX + PUSH_NOTIFICATIONS_RECENTLY_WORKING_REDIS_KEY
+    # Keep this record around for 24h in case it's useful for debugging.
+    redis_client.set(redis_key, str(timezone_now().timestamp()), ex=60 * 60 * 24)
+
+
+def check_push_notifications_recently_working() -> bool:
+    # Check in redis whether push notifications were working in the last hour.
+    redis_key = redis_utils.REDIS_KEY_PREFIX + PUSH_NOTIFICATIONS_RECENTLY_WORKING_REDIS_KEY
+    timestamp = redis_client.get(redis_key)
+    if timestamp is None:
+        return False
+
+    # If the timestamp is within the last hour, we consider push notifications to be working.
+    return timezone_now().timestamp() - float(timestamp) < 60 * 60
+
+
 def maybe_mark_pushes_disabled(
```

Comment:
This is only called from `send_server_data_to_push_bouncer` which is only ever called once an hour.  I'm not clear how this can possibly help, since the "recently working" is only once an hour.

Reply:
The "recently working"  get recorded when a push notification goes through, so that's generally high frequency. So if there's a transient error connecting to the bouncer when the analytics upload job runs (which is indeed once an hour) it will look at "recently working", which will actually likely be a very recent timestamp. And therefore it will stop the job from wrongfully flapping `push_notifications_enabled`

Reply:
A follow-up question that came to my mind here when implementing the issue was "okay, but then aren't these high-frequency push notification attempts going to run into this transient error anyway and flap `push_notifications_enabled`?" But the answer to that is that sending push notifications doesn't flip `push_notifications_enabled` on transient error. Only the upload analytics job does

Reply:
Ah, OK -- for some reason I thought that `record_push_notifications_recently_working` was also in the hourly path, but I see that it's every push.

I'm a little unsettled by having increasingly divergent treatment of errors on the aanalytics codepath from the push codepath, but I don't have a specific failure mode I can name.